File: Leetcode/Easy/two-sum.cpp
/*
https://leetcode.com/problems/two-sum/
*/

#include <iostream>
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        
        unordered_map<int, int> sum_pairs;
        
        for(int i = 0; i < nums.size(); i++) {
            
            if (sum_pairs.find(target - nums[i]) != sum_pairs.end())    
                return {i, sum_pairs[target - nums[i]]};              
                
            else
                sum_pairs[nums[i]] = i;  
            
        }
        return {0, 0};
    }  

    
    vector<int> inputArr() {

        int n, num;
        vector<int> nums;        
        
        cin >> n;
        for(int i = 0; i < n; i++) {
            cin >> num;
            nums.push_back(num);
        }

        return nums;             
    }  
        
};

int main() {

    /*
    4
    2 7 11 15
    9
    */

    /*
    0 1
    */

    Solution *soln = new Solution;

    vector<int> nums = soln -> inputArr();
    int target;
    cin >> target;
    
    vector<int> sum_pair = soln -> twoSum(nums, target);
    
    cout << sum_pair[0] << " " << sum_pair[1] << endl;
    
    return 0;

}

File: Data-Structures-Algorithms/Arrays/Arrays-1D/two-sum.py
# https://leetcode.com/problems/two-sum/

class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        numdict = {}
        for i in range(len(nums)):
            if nums[i] in numdict:
                numdict[nums[i]].append(i)
            else:
                numdict[nums[i]] = [i]
        l = []
        for i in range(len(nums)):
            diff = target - nums[i]
            l.append(diff)
            if diff in numdict:
                if diff != nums[i]:
                    return [i, numdict[diff][0]]
                else:
                    if len(numdict[diff]) > 1:
                        return [numdict[diff][0], numdict[diff][1]]

        return [0,0]


        